### PR TITLE
Add new backpressure algo to replace aimd

### LINKED ISSUE #8298: Introduce a less pessimistic backpressure algorithm
**Is your feature request related to a problem? Please describe.**
Users have quite often complained that they get "RESOURCE_EXHAUSTED" errors too frequent, even if the system is not supposed to be in high load.  The default backpressure algorithm is vegas. The limits calculated by vegas is a bit a pessimistic. Moreover it is also difficult to tune to a specific latency/throughput requirement.

AIMD, another backpressure algorithm available in Zeebe, is easy to configure. You can configure a RequestTimeout and the limit is calculated based on this value. The RequestTimeout is the expected latency of a single request. The limit is maintained in such a way that most of the requests are completed within the configured RequestTimeout. The limit is adjusted based on the observed latency. This sounds ideal. But the implementation of AIMD is not optimal. In high load scenario, aimd limit always fluctuates between minLimit to 2 * optimalLimit. As a result we never get a stable limit and throughput is usually below optimal.

**Describe the solution you'd like**
 Calculating limit based on a configured RequestTimeout is a good idea. This is also easy to configure as there are not many parameters to consider. Since AIMD does not work well, I propose a new algo which fixes some of the issues with the original AIMD.

Updating the limit in AIMD algorithm based on the oberved `rtt` of a request:
```
	if (didDrop || rtt > timeout) {
            currentLimit = (int) (currentLimit * backoffRatio);
        } else if (inflight * 2 >= currentLimit) {
            currentLimit =  currentLimit + 1;
        }
```

Updating the limit in New-aimd algorithm based on the oberved `rtt` of a request:
```
   if ((didDrop || rtt > timeout)) {
      if (inflight <= currentLimit) { // this check ensures that if it is already at the optimal limit, we don't reduce the limit further
        currentLimit = (int) (currentLimit * backoffRatio);
      }
    } else if (inflight * 2 >= currentLimit) {
      currentLimit = currentLimit + 1;
    }
```
This additional check `if (inflight <= currentLimit)` is all that is needed to fix AIMD algo. This check ensures that we don't reduce the limit further if we have already found an optimal limit. Here `inflight` is the number of inflight request at the time this request was sent. Let's say if the inflight was 10, and the rtt > timeout; then it indicates that 10 is too many requests. So we reduce the currentLimit only if it is more than 10. On the other hand, if the inflight was 20 and rtt > timeout, and if the currentLimit is 10, we do not reduce it further because at this point we already know that 20 is too many requests, that is why the currentLimit is 10. But we don't have any indication that 10 is too many, so we leave it as it is. 

**Describe alternatives you've considered**

Completely rethink backpressure approach.

**Additional context**
I ran some benchmark to evaluate to new-aimd implementation.
Note: Only ran with a single configuration/load. The results may change in different settings.

TLDR: 
* New-aimd has less number of request rejected due to backpressure compared to default
* New-aimd has slightly better throughput than default
* New-aimd has **higher process completion latency** compared to default
* It reacts well to change in processing latency => Limit is reduced when there is an increase in processing latency and vice-versa.

Configuration:
```
clusterSize: 3
partitionCount: 1
replicationFactor: 3

-Dapp.starter.rate=100

-Dapp.worker.capacity=50
-Dapp.worker.pollingDelay=1ms
-Dapp.worker.completionDelay=50ms
 worker.replicas = 6
 ```
New-aimd config:
```
  - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_REQUESTTIMEOUT
    value: "200ms"
  - name: ZEEBE_BROKER_BACKPRESSURE_USEWINDOWED
    value: "true"
```

| New-aimd  | Default |
| ------------- | ------------- |
| image | image  |

New-aimd:

Backpressure limit is higher than default. Meaning less requests are rejected due to backpressure. This result in slightly better throughput. Note that the result will be different, if we double the load to stress test the system. 

The avg process completion latency (Process Instance Execution time) is higher. This is because the number of un-processed records in logstreams is higher due to higher backpressure queue limit. 

The overall processing latency (~latency of a single request) stays on avg between 100-250ms, which is the configured RequestTimeout (200ms) for new-aimd. 

To see how it reacts to unexpected fluctuations in the system, we added some stress to the cpu, by running the `stress -c 4 -d 2`.

| New-aimd  | Default |
| ------------- | ------------- |
|image  |  image |

We can observe that new-aimd reacts to the stress quite fast and reduce the limit. The overall processing latency was maintained between 100-500ms, not too distance from the configured RequestTimeout (200ms). 

# Support:
- SUPPORT-15888

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Introduce a new algorithm based on AIMD limit as discused in #8298 
AIMD is replaced with the new one. It uses the same configuration as aimd. When configured algo is aimd, it used the new one instead.

## Related issues

closes #8298

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [x] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by deepthidevaki** [comment]
To eliminate the interference of multiple partition, evaluted the new algo(saimd) with the following configuration:
```
clusterSize:3
partitionCount:1
replicationFactor:3
```
For saimd, following configuration is used while varying the REQUESTTIMEOUT. Evaluated it with different RequestTimeout - 50ms, 100ms, 200ms, 500ms.

```
    - name: ZEEBE_BROKER_BACKPRESSURE_ALGORITHM
      value: "aimd"
    - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_REQUESTTIMEOUT
      value: "50ms"
    - name: ZEEBE_BROKER_BACKPRESSURE_USEWINDOWED
      value: "true"
    - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_INITIALLIMIT
      value: "100"
    - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_MINLIMIT
      value: "1"
    - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_BACKOFFRATIO
      value: "0.9"
```
 For comparison, the default(vegas) is also run using the same cluster configuration.

The limit is higher with higher request timeout. saimd has higher limit than default
image
Backpressure ratio is better for saimd compared to default. But there is no significant different between saimd with different request timeouts, even though limits vary.
image

Since backpressure is higher for default, throughput is also slightly lower compared to saimd.
image

Since the limit is higher, the requests are queued for longer time, resulting in higher process execution duration for saimd with higher request timeouts:
image


Overall processing latency (per request) is also higher with higher limit.
Overall processsing latency: default
image
Overall processing latency: saimd-100
image
Overall processing latency:saimd-500 
image

**Comment by deepthidevaki** [comment]
I think it make sense to make it the default, because it is more intuitive and easier to configure. You only have to configure the requestTimeout. I haven't set this as default yet, because:
1. Anything else to evaluate?
2. It is not yet clear what would be a good default value for the requestTimeout. A higher value (>500ms) would probably reduce the number of resource_exhausted errors observed by someone trying zeebe for the first time. But a higher value would not give you an optimal throughput vs latency. For production, users must evaluate and find a good configuration. Based on our benchmarl, a request timeout of 100-200ms sounds reasonable. But this might be good only for the specific cluster config/infra.

**Comment by deepthidevaki** [comment]
Also running it with our default benchmark setup.

saimd with requestTimeout = 200ms (Following result when leaders are equally distributed over all brokers)
image
image
image

**Comment by deepthidevaki** [comment]
Running with default benchmark setup.

saimd with requestTimeout = 100ms
image

**Comment by oleschoenburg** [comment]
Probably out of scope but in general it'd be cool if we could try out saimd as the log storage appender backpressure algorithm too. Maybe we could move `StabilizingAIMDLimit` to some common module (scheduler, maybe?) so that it is at least available outside of the broker.

**Comment by npepinpe** [comment]
I would assume the request time out to be highly based on workload and usage characteristics. However, we do run a SaaS, and it doesn't scale to tune this for each user (at least not manually). I think the 200ms sounds reasonable for most users, as this is the maximum request timeout under high load, not the general target time. We should make it clear in the release notes that the defaults for this have changed though, so self-managed users should monitor their throughput/latency afterwards in case they need to revert back to the previous defaults (if they so happened to work better for them).

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  static final class Builder {
+    private int minLimit = 10;
+    private int initialLimit = 100;
+    private int maxLimit = 1000;
+    private double backoffRatio = 0.9;
+    private long expectedRtt = DEFAULT_MAX_RTT;
+
+    public StabilizingAIMDLimit.Builder initialLimit(final int initialLimit) {
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:78]: :question: Why are we prefixing the class from within itself?

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * time). The algorithm tries to keep the rtts around expectedRTT. It is not guaranteed to be always
+ * less than that. But on an average, rtts will be less than the configured expectedRTT.
+ *
+ * <p>The implementation is based on {@link com.netflix.concurrency.limits.limit.AIMDLimit}.
+ * AIMDLimit has a limitation that the limit fluctuates between 1 and 2*X, where X is the optimal
+ * limit for the system. It rarely stabilizes at X. This results in always fluctuating throughput
+ * and latency. {@link StabilizingAIMDLimit} fixes this issue and attempts to keep the limit around
+ * X.
+ */
+public final class StabilizingAIMDLimit extends AbstractLimit {
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:25]:
```suggestion
final class StabilizingAIMDLimit extends AbstractLimit {
```

Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:25]: :wrench: Generally if it's only used at the package level, then I would set all visibility to package. I think it's a good rule of them to always assign the weakest visibility possible as a defensive measure - it's easy to make it broader, but harder to weaken again in the future due to misuse or the likes.

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      final long expectedRTT) {
+    super(initialLimit);
+    this.maxLimit = maxLimit;
+    this.minLimit = minLimit;
+    this.backoffRatio = backoffRatio;
+    this.expectedRTT = expectedRTT;
+  }
+
+  @Override
+  protected int _update(
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:47]: :question: Do you think we don't need tests? How can we ensure that changes we do will not break or change the behavior negatively in the future?

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java (hunk 1 above)
Comment (deepthidevaki) [reply broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java]: Fixed.

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java (hunk 3 above)
Comment (deepthidevaki) [reply broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:47]: I have now added tests for it.

**Comment by deepthidevaki** [comment]
> Probably out of scope but in general it'd be cool if we could try out saimd as the log storage appender backpressure algorithm too. Maybe we could move `StabilizingAIMDLimit` to some common module (scheduler, maybe?) so that it is at least available outside of the broker.

We could consider using this. For now, I will leave it here. We can move it a common package, if we find it useful in appender backpressure.

**Comment by deepthidevaki** [comment]
@npepinpe Should we already set the default algo from vegas to aimd? Then, I would also change the default values for aimd config.

**Comment by npepinpe** [comment]
By setting the default you mean to SAIMD, not AIMD right? Then yes, so we can start testing it ASAP :+1:

**Comment by deepthidevaki** [comment]
> By setting the default you mean to SAIMD, not AIMD right? Then yes, so we can start testing it ASAP +1

I have replaced aimd with saimd. So when you configure aimd, it will use saimd.

**Comment by deepthidevaki** [comment]
@npepinpe I have applied your review comments and set the default to aimd.

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java (hunk 3 above)
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:47]: I wasn't sure to request it because I myself wasn't sure how to test this :smile:

File: broker/src/test/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimitTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class StabilizingAIMDLimitTest {
+
+  private StabilizingAIMDLimit limit;
+
+  @BeforeEach
+  void setup() {
+    limit =
```
Comment (npepinpe) [inline broker/src/test/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimitTest.java:22]: :wrench: As this is not dynamic, it could simply go in the declaration above.

```java
priva final StabilizingAIMDLimit limit =         StabilizingAIMDLimit.newBuilder()
            .initialLimit(10)
            .minLimit(1)
            .expectedRTT(1, TimeUnit.SECONDS)
            .build();
```

File: broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * The limit is calculated purely based on the configured expectedRTT and observed rtt (round trip
+ * time). The algorithm tries to keep the rtts around expectedRTT. It is not guaranteed to be always
+ * less than that. But on an average, rtts will be less than the configured expectedRTT.
+ *
+ * <p>The implementation is based on {@link com.netflix.concurrency.limits.limit.AIMDLimit}.
+ * AIMDLimit has a limitation that the limit fluctuates between 1 and 2*X, where X is the optimal
+ * limit for the system. It rarely stabilizes at X. This results in always fluctuating throughput
+ * and latency. {@link StabilizingAIMDLimit} fixes this issue and attempts to keep the limit around
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/transport/backpressure/StabilizingAIMDLimit.java:22]: :wrench: I would add a short one-liner of how it fixes the issue, not just that it does, as the how is the interesting part. Although to me it looks like this is the more "accurate" implementation of AIMD the way I understood it, and I don't quite understand why Netflix's will cut the limit by half if it's reached the max limit (if you know, I'm very curious as to why).

**Review by npepinpe — APPROVED** [review]
:+1:

**Comment by deepthidevaki** [comment]
bors merge
